### PR TITLE
[Feat] File upload

### PR DESCRIPTION
**Pull request by fedeizzo**
What does this PR do?
---

Motivation
---
#3 

Testing
---
1. create a user in the db
2. run `curl -X POST -H "Content-Type: multipart/form-data" -H "Authorization: Bearer <TOKEN>" -F file=@CODEOWNERS http://127.0.0.1:8888/api/files` where the access token is the userID in the db
3. Enjoy the uploaded file 🚀 

```sh
$ llt files
 files
└──  018f3446-3e0b-715b-b1d5-6351f5ac7c93
   └──  CODEOWNERS
```

### CODE REVIEW

File: server/internal/config/storage.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package config
+
+type StorageConfig struct {
+	StorageRoot string `group:"storage" default:"/oxidrive/storage/files"`
```
Comment (MatteoJoliveau) [inline server/internal/config/storage.go:4]:
```suggestion
	StoragePrefix string `group:"storage" default:"/files"`
```

I don't think it makes sense to put this much nesting in the storage root.
This needs to be valid for both FS and S3, and can be combined with a mount point (for FS)/bucket name (for S3) to get the final storage location.

Example with FS:
```
StoragePrefix: /files
StorageFsMountPoint: /oxidrive/storage

Final location: /oxidrive/storage/files
```


Example with S3 (path-style):
```
StoragePrefix: /files
StorageS3Bucket: oxidrive
StorageS3Host: s3.amazonaws.com

Final location: s3.amazonaws.com/oxidrive/files
```

Example with S3 (vhost-style):
```
StoragePrefix: /files
StorageS3Bucket: oxidrive
StorageS3Host: s3.amazonaws.com

Final location: oxidrive.s3.amazonaws.com/files
```

Comment (MatteoJoliveau) [reply server/internal/config/storage.go:4]: This allows Oxidrive to reuse the base location (aka the bucket or mount point) for internal files, like internal backups or profile picture if we ever implement them

File: server/internal/core/application.go [hunk 2]
```diff
@@ -1,24 +1,62 @@
...
 
 func (app *Application) Instance() *instance.Service {
 	return &app.instance
 }
+
+func (app *Application) File() *file.Service {
+	return &app.file
+}
+
+func SetupDependencies(cfg config.Config, db *sqlx.DB) ApplicationDependencies {
```
Comment (MatteoJoliveau) [inline server/internal/core/application.go:40]: We cannot depend on real-word concepts and libraries like sqlx in the `core` module!
The dependencies wiring needs to happen outside, in the main (ideally) or infrastructure layers

File: server/internal/core/file/file.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func isValid(path Path) bool {
+	cleaned := filepath.Clean(string(path))
+
+	return filepath.IsLocal(cleaned)
+}
+
+type FilesContent interface {
+	Store(context.Context, user.User, File, zerolog.Logger) error
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:59]:
```suggestion
	Store(context.Context, File) error
```
Logging is infrastructure, if a `FilesContent` implementation wants to log it can hold its own logger instance.

As for the user, we don't need to pass it. We're persisting the file, not the user. To keep track of the owner, let's add a `ownerID UserID` field in the `File` struct instead

File: server/internal/core/file/file.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+type FilesMetadata interface {
+	Save(context.Context, user.User, File, zerolog.Logger) (*File, error)
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:63]: Same as above

File: server/internal/core/file/file_mock_test.go
Comment (MatteoJoliveau) [inline server/internal/core/file/file_mock_test.go:1]: This test is superfluous, we're basically testing the `stretchr/testify/mock` library which I would hope is already tested in its own codebase.

File: server/internal/core/file/file.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+type File struct {
+	ID      FileID
+	Content Content
+	Name    Name
+	Path    Path
+	Size    Size
+}
+
+func NewFile(content Content, path Path, size Size) (*File, error) {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:36]:
```suggestion
func Create(content Content, path Path, size Size) (*File, error) {
```
Naming nitpick: I prefer more expressive names to "constructor-style" functions. `file.Create` reads better that `file.NewFile`, and `file.New` doesn't convey the feeling of "action" like `Create` does.

File: server/internal/core/file/service.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func NewService(filesContent FilesContent, filesMetadata FilesMetadata) Service {
+	return Service{
+		filesContent:  filesContent,
+		filesMetadata: filesMetadata,
+	}
+}
+
+func (s *Service) Upload(ctx context.Context, u user.User, content Content, path Path, size Size, logger zerolog.Logger) error {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/service.go:23]: Could the content, path, and size params be put in a `FileUpload` struct or something similar? Functions with more than 4 arguments are hard to visually parse

Comment (MatteoJoliveau) [reply server/internal/core/file/service.go:23]: Also, it's better to put the logger instance inside the `Service` struct than to pass it as an arg (and soon we will actually get rid of it completely in favour of domain probes)

File: server/internal/infrastructure/file/metadata_sqlite_test.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+		assert.Error(t, err)
+		assert.Nil(t, saved)
+	})
+
+}
+
+func insertSqliteUser(t *testing.T, db *sqlx.DB, username string, password string) user.User {
+	t.Helper()
+
+	users := userinfra.NewPgUsers(db)
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/metadata_sqlite_test.go:99]:
```suggestion
	users := userinfra.NewSqliteUsers(db)
```

File: server/internal/infrastructure/file/metadata_sqlite_test.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+		assert.Equal(t, fileToSave.Size, saved.Size)
+	})
+
+	t.Run("do not saves with a not existing user", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithPgDB())
+		defer done()
+
+		db := testutil.PgDBFromContext(ctx, t)
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/metadata_sqlite_test.go:81]:
```suggestion
		db := testutil.SqliteDBFromContext(ctx, t)
```

File: server/internal/infrastructure/file/metadata_sqlite_test.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+		assert.NoError(t, err)
+		assert.Equal(t, fileToSave.Name, saved.Name)
+		assert.Equal(t, fileToSave.Path, saved.Path)
+		assert.Equal(t, fileToSave.Size, saved.Size)
+	})
+
+	t.Run("do not saves with a not existing user", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithPgDB())
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/metadata_sqlite_test.go:78]:
```suggestion
		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithSqliteDB(testutil.SqliteDBConfig{}))
```

File: server/internal/infrastructure/user/user_pg_test.go [hunk 10]
```diff
@@ -21,8 +21,8 @@ func TestPgUsers_Count(t *testing.T) {
 
 		users := NewPgUsers(db)
 
-		testutil.Must(users.Save(ctx, testutil.Must(user.Create("a", "a"))))
-		testutil.Must(users.Save(ctx, testutil.Must(user.Create("b", "b"))))
+		testutil.Must(users.Save(ctx, *testutil.Must(user.Create("a", "a"))))
+		testutil.Must(users.Save(ctx, *testutil.Must(user.Create("b", "b"))))
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/user/user_pg_test.go:25]: Is it really needed? `testutil.Must` returns T, so if `user.Create` returns `*User` we're already covered

File: server/internal/web/middleware/authentication.go
Comment (MatteoJoliveau) [inline server/internal/web/middleware/authentication.go:1]: Meh we'll throw it out with #2 , I'd say we can go without authentication for now (just fake it by passing the user ID in a header)

File: server/migrations/postgres/20240422191813_create-files.up.sql [hunk 11]
```diff
@@ -0,0 +1,8 @@
+create table files (
+    id uuid primary key,
+    name text not null,
```
Comment (MatteoJoliveau) [inline server/migrations/postgres/20240422191813_create-files.up.sql:3]: Big question, I don't remember our previous discussion: should this be case sensitive or not?

File: server/migrations/postgres/20240422191813_create-files.up.sql [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+    path text unique not null,
+    size bigint not null,
+    user_id uuid,
+    constraint fk_user_id foreign key(user_id) references users(id) on delete cascade
```
Comment (MatteoJoliveau) [inline server/migrations/postgres/20240422191813_create-files.up.sql:7]: Are we sure we want to delete all files of a user if we delete the user?

Deleting files is nearly always a bad thing, this could lead to dangerous data loss. Maybe we should just orphan the files, then have a manual garbage collection system so that admins can cleanup dead files?

**Review by MatteoJoliveau — CHANGES_REQUESTED** [review]

File: server/internal/config/storage.go (hunk 1 above)
Comment (fedeizzo) [reply server/internal/config/storage.go:4]: 👍

File: server/internal/core/file/file.go (hunk 5 above)
Comment (fedeizzo) [reply server/internal/core/file/file.go:36]: `Create` is ambiguous, multiple structs (file and service in this case) are defined in the same package

File: server/internal/infrastructure/file/metadata_sqlite_test.go (hunk 7 above)
Comment (fedeizzo) [reply server/internal/infrastructure/file/metadata_sqlite_test.go:99]: thx, copy&paste errorr

File: server/internal/web/middleware/authentication.go
Comment (fedeizzo) [reply server/internal/web/middleware/authentication.go:1]: I agree, this also solves the problem of retrieving the id given the username+pwd combo

File: server/migrations/postgres/20240422191813_create-files.up.sql (hunk 11 above)
Comment (fedeizzo) [reply server/migrations/postgres/20240422191813_create-files.up.sql:3]: Right, I completely forgot about it.

Lower case for the comparison but everything is saved as provided by the user.

```
this/is/my/directory/tex.txt
// equal to
This/IS/mY/directory/tExT.txt
```

but in the UI it should be displayed as `This/IS/mY/directory/tExT.txt`.

It can be misleading for the user but it solves several problems with mac os and windows.

File: server/migrations/postgres/20240422191813_create-files.up.sql (hunk 12 above)
Comment (fedeizzo) [reply server/migrations/postgres/20240422191813_create-files.up.sql:7]: No strong opinion on this, if you think that being a soft-delete is a better option and a possible feature, I can remove the delete on cascade

File: server/internal/core/application.go (hunk 2 above)
Comment (fedeizzo) [reply server/internal/core/application.go:40]: I see your point, but I would keep it outside of the `main` file, if the complexity grows I want to test the function that initializes the dependencies and testing a binary is horrible. We can move it inside the infra layer and use a decorator or something similar for the setup. wdyt?

File: server/migrations/postgres/20240422191813_create-files.up.sql (hunk 11 above)
Comment (MatteoJoliveau) [reply server/migrations/postgres/20240422191813_create-files.up.sql:3]: Ok, then I would use `citext` as the column type, because

> Essentially, it internally calls lower when comparing values. Otherwise, it behaves almost exactly like text.

Cit https://www.postgresql.org/docs/current/citext.html

File: server/internal/core/file/file.go (hunk 5 above)
Comment (MatteoJoliveau) [reply server/internal/core/file/file.go:36]: Even if the module name is the same as the struct? It's pretty unambiguous that `file.Create` creates a file, while `file.NewService` initializes a new FileService
Ultimately it's no big deal :)

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (fedeizzo) [reply server/internal/infrastructure/user/user_pg_test.go:25]: What do you mean?

File: server/internal/core/file/service.go (hunk 6 above)
Comment (fedeizzo) [reply server/internal/core/file/service.go:23]: If we remove the `user` and the `logger`, arguments are reduced to three (ctx does not count since it's basically in any function in Go 🙃 ). I would keep them as arguments and not as an additional struct to avoid to jump back and forth only to understand what the function needs

File: server/internal/core/file/file.go (hunk 5 above)
Comment (fedeizzo) [reply server/internal/core/file/file.go:36]: Good point 🤔 

What about this?
```
struct File { ... }
struct FileService { ... }

func Create() File { ... }
func InitService() File { ... }
```
and the final result when you use the `file` package would be
```
import ".../.../.../file"

uploaded := file.Create(...)
fileService := file.InitService(...)
fileService.Upload(uploaded)
```

File: server/internal/web/handler/decoder.go [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+	if !errors.Is(err, io.EOF) {
+		msg := "Request body must only contain a single JSON object"
+		return &MalformedRequest{Status: http.StatusBadRequest, Msg: msg}
+	}
+
+	return nil
+}
+
+func DecodeMutipartBody(w http.ResponseWriter, r *http.Request, dst *MultipartRequest, logger zerolog.Logger) error {
+	if err := r.ParseMultipartForm(0); err != nil {
```
Comment (fedeizzo) [inline server/internal/web/handler/decoder.go:66]: @MatteoJoliveau When you parse the form you can specify the maximum size of the memory buffer, if the file is bigger than the buffer the remaining content is saved temporary on disk.

Given the scope of Oxidrive I expect people to upload big files that potentially are larger than the memory. For this reason I set the value to zero to reduce the footprint in memory.

Any opinion on this?

File: server/internal/infrastructure/file/blob_fs.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+	"github.com/oxidrive/oxidrive/server/internal/config"
+	"github.com/oxidrive/oxidrive/server/internal/core/file"
+	"github.com/oxidrive/oxidrive/server/internal/core/user"
+
+	"github.com/rs/zerolog"
+)
+
+const (
+	throughputInByte    = 32
```
Comment (fedeizzo) [inline server/internal/infrastructure/file/blob_fs.go:19]: Similar consideration for this

File: server/internal/core/file/file_mock_test.go
Comment (MatteoJoliveau) [reply server/internal/core/file/file_mock_test.go:1]: Mmmh I see your point, although I'm a bit worried about this because it adds workload on top of developing the "real" code :c

I'm inclined to say that regular tests should exercise the mock structure as well and thus count as testing it, but if you feel it avoids issues down the line I'm ok with keeping it

File: server/internal/core/file/file.go (hunk 5 above)
Comment (MatteoJoliveau) [reply server/internal/core/file/file.go:36]: Yeah that works. I'm not hot on `Init` vs `New` for the service but I'm not against either, as long as we keep all services similar. If you decide to go for `InitService` please rename the UserService as well

File: server/internal/core/file/service.go (hunk 6 above)
Comment (MatteoJoliveau) [reply server/internal/core/file/service.go:23]: We do need the user here to know who is the owner of the file tho

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (MatteoJoliveau) [reply server/internal/infrastructure/user/user_pg_test.go:25]: You added a pointer reference on the return type of `testutil.Must`, however `Must` is generic over `T` and returns `T`.

Unless generics in Go work wildly differently compared to other languages, that means that passing `*user.User` to `Must` returns ` *user.User`, right?

In that case you don't need to add another `*`

File: server/internal/infrastructure/file/blob_fs.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+	for {
+		if err = ctx.Err(); err != nil {
+			if reErr := os.Remove(string(fsPath)); reErr != nil {
+				logger.Error().AnErr("error", reErr).Msg("error while removing the file in blob fs store after context invalidation")
+			}
+			return fmt.Errorf("context invalidated while saving the file in blob fs: %w", err)
+		}
+
+		if _, err = io.CopyN(fsFile, f.Content, throughputInByte); err != nil {
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs.go:58]: genuine question: why multiple `io.CopyN` over a single `io.Copy`?

File: server/internal/infrastructure/file/blob_fs_test.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/oxidrive/oxidrive/server/internal/core/file"
+	"github.com/oxidrive/oxidrive/server/internal/core/user"
+	"github.com/oxidrive/oxidrive/server/internal/testutil"
+)
+
+func TestBlobFS_Store(t *testing.T) {
+	contentStr := "This is a test!"
+	content := strings.NewReader(contentStr)
+	u, _ := user.Create("username", "pwd")
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs_test.go:22]:
```suggestion
	u, err := user.Create("username", "pwd")
	require.NoError(t, err)
```

File: server/internal/infrastructure/file/blob_fs_test.go [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+	t.Run("stores a file", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithTempDir())
+		defer done()
+
+		path := testutil.TempDirFromContext(ctx, t)
+		blob := blobFS{filesRoot: path}
+		f, _ := file.NewFile(content, "this/dir/without_error.txt", file.Size(len([]byte(contentStr))))
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs_test.go:32]:
```suggestion
		f, err := file.NewFile(content, "this/dir/without_error.txt", file.Size(len([]byte(contentStr))))
		require.NoError(t, err)
```

File: server/internal/infrastructure/file/blob_fs_test.go [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+	t.Run("stores a file with timedouted context", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithTempDir())
+		defer done()
+
+		path := testutil.TempDirFromContext(ctx, t)
+		blob := blobFS{filesRoot: path}
+		f, _ := file.NewFile(content, "this/dir/timeout_error.txt", file.Size(len([]byte(contentStr))))
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs_test.go:48]:
```suggestion
		f, err := file.NewFile(content, "this/dir/timeout_error.txt", file.Size(len([]byte(contentStr))))
		require.NoError(t, err)
```

File: server/internal/infrastructure/file/blob_fs_test.go [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+	t.Run("stores a file with cancelled context", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, done := testutil.IntegrationTest(context.Background(), t, testutil.WithTempDir())
+		defer done()
+
+		path := testutil.TempDirFromContext(ctx, t)
+		blob := blobFS{filesRoot: path}
+		f, _ := file.NewFile(content, "this/dir/ctx_cancelled_error.txt", file.Size(len([]byte(contentStr))))
```
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs_test.go:67]:
```suggestion
		f, err := file.NewFile(content, "this/dir/ctx_cancelled_error.txt", file.Size(len([]byte(contentStr))))
		require.NoError(t, err)
```

**Review by MatteoJoliveau — CHANGES_REQUESTED** [review]

File: server/internal/core/file/file_mock_test.go
Comment (fedeizzo) [reply server/internal/core/file/file_mock_test.go:1]: I agree, with are small team with the same timezone, we can live without mock test

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (fedeizzo) [reply server/internal/infrastructure/user/user_pg_test.go:25]: `user.Create` returns `*User` and Must returns the same type. Save instead wants `User` so we need to use `*` again to point the content of the pointer

File: server/internal/infrastructure/file/blob_fs.go (hunk 15 above)
Comment (fedeizzo) [reply server/internal/infrastructure/file/blob_fs.go:58]: It gives us more control on the loop, and we can check the context

File: server/internal/infrastructure/infrastructure.go [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+	"github.com/oxidrive/oxidrive/server/internal/config"
+	"github.com/oxidrive/oxidrive/server/internal/core"
+	"github.com/oxidrive/oxidrive/server/internal/core/file"
+	"github.com/oxidrive/oxidrive/server/internal/core/user"
+	fileinfra "github.com/oxidrive/oxidrive/server/internal/infrastructure/file"
+	userinfra "github.com/oxidrive/oxidrive/server/internal/infrastructure/user"
+)
+
+func Setup(cfg config.Config, db *sqlx.DB, logger zerolog.Logger) core.ApplicationDependencies {
```
Comment (fedeizzo) [inline server/internal/infrastructure/infrastructure.go:15]: This can be refactored but I would do it in another PR.

File: server/internal/web/server.go [hunk 21]
```diff
@@ -6,13 +6,15 @@ import (
...
 	"github.com/oxidrive/oxidrive/server/internal/core"
+	"github.com/oxidrive/oxidrive/server/internal/web/handler"
 )
 
 type Config struct {
 	Address        string
 	Application    *core.Application
 	Logger         zerolog.Logger
 	FrontendFolder string
+	handler.Config
```
Comment (fedeizzo) [inline server/internal/web/server.go:17]: I am not sure about the name `handler.Config`

File: server/internal/web/handler/decoder.go [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/rs/zerolog"
+)
+
+func DecodeJSONBody(w http.ResponseWriter, r *http.Request, dst interface{}) error {
```
Comment (fedeizzo) [inline server/internal/web/handler/decoder.go:14]: This handler (and all other handlers and middlewares) is not tested, I think this is not the right moment to do it due to the complexity of the PR. If we want to merge only tested code I can create a new branch based on the current one and merge this PR only when the next one is ready, otherwise I can open a new issue and we can tackle the problem later.

File: server/internal/core/file/file.go [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+		Content: content,
+		Name:    name,
+		Path:    path,
+		Size:    size,
+		OwnerID: ownerID,
+	}, nil
+}
+
+func isValid(path Path) bool {
+	cleaned := filepath.Clean(string(path))
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:55]: beware that `filepath` functions are os-aware and will replace `/` with `\` on Windows

I think we should keep the domain ox-agnostic and use `/` for our paths, then convert them to the os-specific variant in the FS implementation (S3 doesn't care for example, and it can keep using `/` on all platforms).

File: server/internal/core/file/file.go [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+
+	return filepath.IsLocal(cleaned)
+}
+
+type FilesContent interface {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:60]:
```suggestion
type Contents interface {
```
naming nitpick

File: server/internal/core/file/file.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+	Store(context.Context, File) error
+}
+
+type FilesMetadata interface {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file.go:64]:
```suggestion
type Files interface {
```
naming nitpick. I wanted to call it `Metadata` but it would be confusing if we ever add a struct to aggregate file metadata together.

File: server/internal/core/file/file_test.go [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+		file, err := Create(nil, "this/is/a/directory/filename.txt", 5, user.ID(testutil.Must(uuid.NewV7())))
+
+		assert.NotNil(t, file)
+		assert.NoError(t, err)
+	})
+
+	t.Run("returns an error with an invalid path", func(t *testing.T) {
+		t.Parallel()
+
+		file, err := Create(nil, "/this/is/a/directory/filename.txt", 5, user.ID(testutil.Must(uuid.NewV7())))
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file_test.go:27]: Is this really invalid? I know technically it's an absolute path, but it could simply be joined with our mountpoint and prefix right?

File: server/internal/core/file/file_test.go [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+
+		assert.Nil(t, file)
+		assert.ErrorIs(t, err, ErrInvalidPath)
+	})
+}
+
+func TestFile_isValid(t *testing.T) {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/file_test.go:34]: Testing a private function is not recommended, what we need to ensure is working is the public API (aka public methods in this case).

I think these test cases can be simply put in the `Create` suite, checking `assert.ErrorIs(t, err, testCase.expected)`

File: server/internal/core/file/service.go [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func InitService(filesContent FilesContent, filesMetadata FilesMetadata) Service {
+	return Service{
+		filesContent:  filesContent,
+		filesMetadata: filesMetadata,
+	}
+}
+
+func (s *Service) Upload(ctx context.Context, toUpload FileUpload, owner user.ID) error {
```
Comment (MatteoJoliveau) [inline server/internal/core/file/service.go:28]:
```suggestion
func (s *Service) Upload(ctx context.Context, upload FileUpload, owner user.ID) error {
```
Naming nitpick

File: server/internal/core/file/service.go [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+	// TODO add user validation logic
+	f, err := Create(toUpload.Content, toUpload.Path, toUpload.Size, owner)
+	if err != nil {
+		return err
```
Comment (MatteoJoliveau) [inline server/internal/core/file/service.go:32]:
```suggestion
		return fmt.Errorf("failed to create new file from upload: %w", err)
```

File: server/internal/infrastructure/file/blob_fs.go
Comment (MatteoJoliveau) [inline server/internal/infrastructure/file/blob_fs.go:1]: Since we decided for Content instead of Blob (which I quite like), let's call this `content_fs`/ `contentFs`

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (MatteoJoliveau) [reply server/internal/infrastructure/user/user_pg_test.go:25]: Aaaaaah right I got it.
Would it make sense for `Save` to then accept a `*user.User`? Which one is more idiomatic?

File: server/internal/testutil/sqlite.go [hunk 30]
```diff
@@ -78,5 +79,10 @@ func SqliteDBFromContext(ctx context.Context, t *testing.T) *sqlx.DB {
 		t.Fatal(err)
 	}
 
+	_, err = db.Exec("PRAGMA foreign_keys = ON;")
```
Comment (MatteoJoliveau) [inline server/internal/testutil/sqlite.go:82]: Could we use `infrastructure.InitDB` instead to avoid duplication?

File: server/internal/web/handler/decoder.go (hunk 22 above)
Comment (MatteoJoliveau) [reply server/internal/web/handler/decoder.go:14]: I don't think I would unit-test handlers, we have E2E tests to cover the API as well. Unless we have specific concerns or failure scenarios we want to enforce, Handlers and middlewares are pretty dumb as all the meaty logic is in the domain. Once that is tested, we can reasonably assume they will work as well.

File: server/internal/web/server.go (hunk 21 above)
Comment (MatteoJoliveau) [reply server/internal/web/server.go:17]: Why not?

**Review by MatteoJoliveau — CHANGES_REQUESTED** [review]

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (fedeizzo) [reply server/internal/infrastructure/user/user_pg_test.go:25]: So, if the struct contains a lot of data you want to use the pointer to prevent it to be copied every time, in this case the user size is limited, so we can pass the argument without a pointer.

I personally prefer to pass always a struct without a pointer to avoid possible bugs if no validation is applied on `nil`, the idea is to move up in the stack trace the validation `if user != nil then use *user` but of course it has a cost in memory (negligible this time).

Up to you, for me it's the same

File: server/internal/web/server.go (hunk 21 above)
Comment (fedeizzo) [reply server/internal/web/server.go:17]: Maybe it's more server config, rather than handler config?

Comment (fedeizzo) [reply server/internal/web/server.go:17]: It creates multiple configs scattered in different places, and it may be hard to have a quick look at them

File: server/internal/core/file/file.go (hunk 23 above)
Comment (fedeizzo) [reply server/internal/core/file/file.go:55]: 💩

File: server/internal/core/file/file_test.go (hunk 26 above)
Comment (fedeizzo) [reply server/internal/core/file/file_test.go:27]: If we decide to accept absolute paths, there are two consequences:
- `/this/is/a/directory/filename.txt` will be equal to `this/is/a/directory/filename.txt`
- It adds a layer of complexity in the validation, we have to check for this `/this/../../../../filename.txt`, which is not necessary complex

It could be a good idea, right now with the API you cannot upload a file using an absolute path, unless you modify the header of the form-data
```sh
curl -X POST -H "Content-Type: multipart/form-data" -H "Authorization: Bearer <TOKEN>" -F file=@/root/CODEOWNERS http://127.0.0.1:8888/api/files
```

WDYT?

File: server/internal/testutil/sqlite.go (hunk 30 above)
Comment (fedeizzo) [reply server/internal/testutil/sqlite.go:82]: Unfortunately no. If you try to use `infrastructure.InitDB` you have a circular dependency.
```
package github.com/oxidrive/oxidrive/server/internal/infrastructure/file
	imports github.com/oxidrive/oxidrive/server/internal/testutil
	imports github.com/oxidrive/oxidrive/server/internal/infrastructure
	imports github.com/oxidrive/oxidrive/server/internal/infrastructure/file: import cycle not allowed in test
```

File: server/internal/core/file/service.go (hunk 29 above)
Comment (MatteoJoliveau) [reply server/internal/core/file/service.go:32]: Nevermind then, discarding this

File: server/internal/core/file/file_test.go (hunk 26 above)
Comment (MatteoJoliveau) [reply server/internal/core/file/file_test.go:27]: Maybe we could simply normalize the path by adding an initial `/` if it's not already there?

File: server/internal/infrastructure/user/user_pg_test.go (hunk 10 above)
Comment (MatteoJoliveau) [reply server/internal/infrastructure/user/user_pg_test.go:25]: Let's keep it like it is then

**Review by MatteoJoliveau — APPROVED** [review]
